### ISSUE TITLE
[77] DeepNet: Scaling Transformers to 1,000 Layers

### ISSUE THREAD

**Issue by dhkim0225** [dhkim0225/1day_1paper#106]
[paper](https://arxiv.org/abs/2203.00555)
[code](https://github.com/microsoft/unilm)

`rosinality`'s comment
```
pre norm의 안정성과 post norm의 성능. 
post norm을 사용하는 경우 학습 초반에 큰 크기의 업데이트가 발생하고 이 업데이트가 layer norm과 맞물려 gradient vanishing이 발생

학습 초반에 너무 큰 업데이트가 발생하는 것이 문제라는 게 warmup의 문제의식.
학습 초반의 업데이트가 폭발하지 않도록 그 크기를 제한하도록 했음. 

결과적으로 `layernorm(x * a + f(x) * b)` 형태에서 a를 키우고 b의 크기를 줄이는, 즉 skip connection의 크기를 키우는 형태가 됨. 
zero init 같은 걸 생각해보면 친숙한 결과
이걸 써서 모델을 1000 레이어까지 확장할 수 있음. 
겸사겸사 width를 키우는 것보다는 depth를 키우는 것이 더 나은 scaling을 보여준다는 것까지 보였음.
```

![image](https://user-images.githubusercontent.com/16400591/156873186-2683cfa3-c6e0-46d5-9252-7709ec30ea10.png)

## TL;DR for Practitioners
N-layer, M-layer 에 대한 방법.
베타값은 0~1 로 설정된다.
알파값은 키워서 skip con 이 많이 사용되도록 한다.
![image](https://user-images.githubusercontent.com/16400591/157557284-d65ad030-1ee6-4bf8-a142-803c2401100c.png)

layer 많이 쌓을 수록 0에 가깝게 init 해야 안정적임.
![image](https://user-images.githubusercontent.com/16400591/157557959-7e2cba4a-15d2-4808-aa56-6fa12f6bfd30.png)

참고로 pytorch 에서 추천하는 gain 값은 다음과 같음
![image](https://user-images.githubusercontent.com/16400591/157557788-4634f686-88ce-4fe1-8597-1288827b9a86.png)

저렇게 initialize 를 하고 나서는 layer 별로 크기를 줄여 주었다.
![image](https://user-images.githubusercontent.com/16400591/157561286-f96e2406-a989-45e5-af88-a8bf11bd42c3.png)
d` 은 fan_in, fan_out 들의 평균.
l 은 layer num
![image](https://user-images.githubusercontent.com/16400591/157561543-cccaf355-614e-4c72-bd90-85f80b9c652b.png)
k_l 은 단순 scalar 값으로 normalize factor. 뒷 레이어로 갈 수록 값이 작아짐 (마지막 layer는 1)
즉, 뒷 레이어일수록 0에 값이 몰림.

encoder-only 와 encoder-decoder 에 하는 방식 각각 다름
![image](https://user-images.githubusercontent.com/16400591/157563547-4c36e53c-09c4-4164-9a4c-eace03ff0170.png)
![image](https://user-images.githubusercontent.com/16400591/157563562-aa743bcd-c179-435a-967f-c18ed7f11677.png)

## Instability of Deep Transformer
IWSLT-14 De-En machine translation 에 대해서 enc 18L dec 18L  로 실험해 봤음.
![image](https://user-images.githubusercontent.com/16400591/157561080-30153b19-d7a2-4cb9-8b5c-6e177ee2d844.png)

(a) post-ln init 을 쓰면, top layer 일수록 gradient 가 커진다.
(b) post-ln init 을 쓰면, layer 가 깊어질 수록 top layer 는 gradient 가 감소함
(c) post-ln init 을 쓰면, val loss 안정적으로 떨어짐.

![image](https://user-images.githubusercontent.com/16400591/157562951-b0a49689-4110-4f07-89f4-5cb991f1ebf5.png)

(a), (b), (c) 를 통해 warmup 은 반드시 필요하다는 것을 알 수 있다.
근데, 기존 post-LN 에 warmup을 넣어버리면 앞쪽 layer update 와 뒤쪽 layer 업데이트 균형이 참 안 맞다.
post-LN init 을 사용하면 이런 불균형이 해결된다.

## Results
확실히 pre-LN 은 안정적이다. 레이어 쌓아도 결과는 나옴.
![image](https://user-images.githubusercontent.com/16400591/157563387-e0f3d20d-3e49-401b-91f1-2006ea6f92b7.png)
